mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,199 @@
+
+    def load_kube_config(self):
+        config.load_kube_config(
+            config_file=self.kube_config_file_path
+            or os.getenv("KUBECONFIG", default="~/.kube/config")
+        )
+
+    def _get_client(self, state: K8sDeploymentState) -> Client:
+        self.load_kube_config()
+        service = client.CoreV1Api().list_namespaced_service(
+            f"mlem-{state.deployment_name}-app"
+        )
+        port = service.items[0].spec.ports[0].port
+        host = service.items[0].status.load_balancer.ingress[0].ip
+        return HTTPClient(host=host, port=port)
+
+
+class K8sEnv(MlemEnv[K8sDeployment]):
+    type: ClassVar = "kubernetes"
+    deploy_type: ClassVar = K8sDeployment
+    registry: Optional[DockerRegistry] = DockerRegistry()
+
+    def create_k8s_resources(self, generator):
+        k8s_client = client.ApiClient()
+        with tempfile.TemporaryDirectory(
+            prefix="mlem_k8s_yaml_build_"
+        ) as tempdir:
+            filename = os.path.join(tempdir, "resource.yaml")
+            generator.write(filename)
+            utils.create_from_yaml(k8s_client, filename, verbose=True)
+
+    def deploy(self, meta: K8sDeployment):
+        self.check_type(meta)
+        redeploy = False
+        with meta.lock_state():
+            meta.load_kube_config()
+            state: K8sDeploymentState = meta.get_state()
+            if state.image is None or meta.model_changed():
+
+                image_name = meta.name or generate_docker_container_name()
+                echo(EMOJI_BUILD + f"Creating docker image {image_name}")
+                with set_offset(2):
+                    state.image = build_model_image(
+                        meta.get_model(),
+                        image_name,
+                        meta.server
+                        or project_config(
+                            meta.loc.project if meta.is_saved else None
+                        ).server,
+                        DockerEnv(registry=self.registry),
+                        force_overwrite=True,
+                    )
+                meta.update_model_hash(state=state)
+                redeploy = True
+
+            if state.deployment_name is None or redeploy:
+                k8s_yaml_builder = K8sYamlBuilder(
+                    image=state.image,
+                    image_pull_policy=meta.image_pull_policy,
+                    port=meta.port,
+                    service_type=meta.service_type,
+                )
+                generator = K8sYamlGenerator(
+                    **k8s_yaml_builder.get_resource_args().dict()
+                )
+                self.create_k8s_resources(generator)
+
+                sleep(0.5)
+
+                deployments_list = (
+                    client.AppsV1Api().list_namespaced_deployment(
+                        namespace=f"mlem-{meta.name}-app"
+                    )
+                )
+
+                state.deployment_name = deployments_list.items[0].metadata.name
```

Comment:
what if it fails? or what if 0.5 seconds is not enough?
ideally, you should check if pod exists and running or throw error otherwise

Reply:
fixed, we have a watcher now

Reply:
please also check that `deployments_list.items` is not empty

Reply:
done